squeal-postgresql/src/Squeal/PostgreSQL/Binary.hs

```diff
@@ -553,3 +652,34 @@ newtype Only x = Only { fromOnly :: x }
   deriving (Functor,Foldable,Traversable,Eq,Ord,Read,Show,GHC.Generic)
 instance Generic (Only x)
 instance HasDatatypeInfo (Only x)
+
+class HomogeneousProduct
+  (n :: Nat) (x :: Type) (product :: Type)
+  | product -> n where
+    foldlN :: (b -> x -> b) -> b -> product -> b
+    replicateMN :: Monad m => m x -> m product
+instance
+  ( IsProductType product xs
+  , All ((~) x) xs
+  , SListI xs
+  , Length xs ~ n )
+  => HomogeneousProduct n x product where
+
+    foldlN f z =
+      let
+        foldlNP
+          :: All ((~) x') xs'
+          => (z -> x' -> z) -> z -> NP I xs' -> z
+        foldlNP f' z' = \case
+          Nil -> z'
+          I x' :* xs' -> let z'' = f' z' x' in seq z'' $ foldlNP f' z'' xs'
```

Comment:
Ah! Recursion! It burns! The recursive-class-inlining-trick-thing works only calling a class function as the recursive part. Like (totally untested):

```haskell
class FoldlNP l where
  foldlNP :: All ((~) a) l => (b -> a -> b) -> b -> NP I l -> b

instance FoldlNP '[] where
  {-# INLINE foldlNP #-}
  foldlNP f !z Nil = z

instance FoldlNP (x ': xs) where
  {-# INLINE foldlNP #-}
  foldlNP f !z (I x :* xs) = foldlNP f (f z x) xs
```

I opened https://github.com/well-typed/generics-sop/issues/74 as the relevant issue for generics-sop

Reply:
Given the response over at `generics-sop`, I think it would be best to first get a benchmark together for Squeal's binary instances before we worry too much about recursive-class-inlining. We can base it on [this](https://github.com/nikita-volkov/postgresql-binary/blob/master/encoding/Main.hs) and [this](https://github.com/nikita-volkov/postgresql-binary/blob/master/decoding/Main.hs). I've removed the class for now.